### ISSUE TITLE
theora uses inline assembly, can not build godot with msvc 2013 without theora=no

### ISSUE THREAD

**Issue by Griefchief** [godotengine/godot#3098]
**Bugsquad note:** *There seems to be a confusion of which Visual Studio prompt to use when compiling the engine. The documentation needs to be more explicit about this subject.*

my env:

Intel Q6600 msvc 2013 ... windows 7 sp1 x64

As the title states, theora uses inline assembly, so scons + msvc 2013 x64 fails with: 

drivers\theora\cpu.c(59) : error C4235: nonstandard extension used : '_asm' keyword not supported on this architecture

I read on the net that msvc x64 doesn't support inline assembly... that would probably be the reason.

But, 32 bit also fails with theora in even though supposedly __asm is supported in it (not sure why, but it does)...

So, as far as i can tell, this isn't an issue with my compiler/computer, as I can build without theora x64 or x86... What i mean is, the issue is with every msvc (2013, and previous, and maybe 2015) compiler probably...

So, the compiling godot for windows tutorial on the wiki now (probably) won't work for anybody using msvc on x64... because of theora? Am i correct?

scons p=windows   //fails, above error (this should be x64 on my x64win, maybe i'm wrong?)
scons p=windows theora=no //successfull build
scons p=windows bits=32 //fails (not sure why, supposedly _asm is supported in 32bit but maybe not?)
scons p=windows bits=32 theora=no // build succeeds

**Comment by Griefchief** [comment]
After reading a little bit it seems 32 bit msvc suppots __asm (not one _ but two __)... So, that's the reason it fails, among other things... If you google theora and asm you might find this disscution about 6 years ago:

http://lists.xiph.org/pipermail/theora-dev/2009-October/003996.html

It states that it works with gcc x64 but the person is not sure about msvc (I'm far from an expert, but it seems to me it doesn't)...

I have to leave you folks to see what you can do about this, this is where my knowledge of asm magic ends.

**Comment by Griefchief** [comment]
64 bit now compiles, but 32 bit still does not.

Please reopen this issue until these new commits are pushed to master, I am guessing they contain code that disabled assembly for 32bit compile as well?

Thank you.

**Comment by Griefchief** [comment]
Basically, but i believe that MSVC supports only __asm, not _asm... __asm is probably some Microsoft assembly thingy...

https://msdn.microsoft.com/en-us/library/45yd4tzz.aspx

Note how it says "Microsoft Specific" in the upper left corner... might be, unlike in gcc which names _asm three different ways for namspace conflict avoidance purposes... Sorry I am not of more use here, not 100% sure...

Also, if third post here means anything to you:

http://www.vbforums.com/showthread.php?492227-difference-between-_asm-and-__asm

**Comment by punto-** [comment]
Does __asm work on all versions of msvc, or only the latest one? Why are
those thora files using _asm? I assume it was because it worked at some
point..

**Comment by Griefchief** [comment]
All versions, but it's compiler specific asm (hence double underscore)... and theora doesn't support support microsofts assembly or Juan would have switched to that version of theora for msvc... 64 bit build that successed currently is based on theora without assembly as juans commit desc states... He just needs to turn it of for 32bit as well...

Read the second post, those links for more details... but in any case, theora folks don't seem to be in the bussiness of supporting msvc when it comes to assembly...

**Comment by reduz** [comment]
i think this is fixed, test HEAD again, assembly is disabled on 64 bits, as
it's not supported

**Comment by Griefchief** [comment]
32 bit fails now @reduz  64bit works. just put that line to false, should make it work :)

**Comment by Griefchief** [comment]
Also, @reduz, two people above made commits in their own forks of godot referencing this issue... I'm guessing they also changed that line... you might want to check out what they did, i can't figure out how to...

**Comment by akien-mga** [comment]
>  two people above made commits in their own forks of godot referencing this issue... I'm guessing they also changed that line... you might want to check out what they did, i can't figure out how to...

Nope that's just GitHub noise of people who rebased this commit by mistake in their fork (which is pretty easy to do with the cryptic git commands), nothing relevant here :)

**Comment by Griefchief** [comment]
I see @akien-mga, thanks for letting me know... To recap, here is the current status on master that I built just right now with a new checkout (even though i checkout master it says HEAD, and i do not "detach the checkout" in GUI... i am guessing it is the head on github):

1) 32bit fails (it uses theora with _asm, "not supported on this arhitecture")
2) 64 bit build succedes because it uses non assembly theora version (as Juans b21ce6c fix states)

@reduz Here is some more info for you, recaping my reasearch into this issue:

2)My suggestion unless you know something i do not at this point(I am not an expert), is to also disable assemly for theora for 32 bits as well... and here is why:

1) Seems like __asm (double underscore) is meant for compiler specific versions of asm... and _asm is non-standards-compliant asm that is (probably) not supported at all by any version of MSVC... microsofts way of "fixing" assembly for MSVC (x64, but also probably x32) is to use "intrinsics" that get transcompiled into propper assembly but theora folks don't like it because it doesn't produce good asm... so they don't seem to support asm on MSVC (maybe this changed from 2009 but if you didn't see msvc version of theora when you pulled it from their website/repo, probably not)... I tried to find asm (supposedly this is standards compliant asm) or _asm (non standards compliant asm) docs on MSDN produced nothing... ti seems microsoft only supports it's own version of asm, hence double underscore

3)Also, Juan, you are awesome!

**Comment by vnen** [comment]
I would reopen, but I'm actually able to build 32-bits version with VS2013 just fine, with last HEAD (765ffc9).

**Comment by Griefchief** [comment]
Well, I was testing on the wrong head in my last comment, i failed to fetch, only fetched this morning (now i know the exact difference between those two)... thanks for teaching me that...

However... My vc2013 still complains about _asm... I still can't compile, even (765ffc9) (and yes, i'm sure it is 765 this time... are you doing something aditional except scons p=windows bits 32, like that .bat thing added by ex in (https://github.com/godotengine/godot/pull/3322)?

Again, i can build x64...

What gives? :)

**Comment by vnen** [comment]
I'm not sure what's going on. I just run `scons -j4 p=windows bits=32` from the Developer Command Prompt. It builds with no error. There's a simple guide to compile that might help you: https://gist.github.com/ZuBsPaCe/0e1c31960e84df33e20f

Maybe it's related to the Visual Studio version? I have Professional version, but is also builds ok with the Community version of VS 2015.

**Comment by punto-** [comment]
Theora provides versions of their assembler files for msvc, but they use
_asm instead of __asm, those fail both on 64 bits and 32. What I'm asking
is, what is the point of those theora files? they come with the theora
library, they must have worked at some point. If we just change everything
to __asm, will it break on older versions of msvc comiler?

**Comment by Griefchief** [comment]
@vnen I am not really sure either... You said it's 2013 pro... maybe somehow MASM is turning on for you and maybe compiling it on 32bit... or something... maybe theora has that in it's preprocesors or whatever, however MASM turns on if it does... I am just pure guessing... reading about it on this link, it seems that MASM might like AMD... Are you on an AMD processor per chance (Intel)? Probably not relevant but can't hurt to ask...

It's probably the diff between comunity and pro version of 2013... whatever it is... seems @punto-  can't build either? on what msvc version?

https://en.wikipedia.org/wiki/Microsoft_Macro_Assembler

**Comment by Griefchief** [comment]
@vnen Ah, forgot to ask, can you go to  platform/windows/detect.py and turn 

env["x86_opt_vc"]=env["bits"]!="64"

to

env["x86_opt_vc"]=True

and build both 32 and 64 as a test (at least 64 should fail)... I tried it, fails, predictably... can't hurt on your side as well ... I'll see if i can learn more about msvc and theora...

Thank you.

**Comment by reduz** [comment]
if you turn it to False it works in both cases?

**Comment by Griefchief** [comment]
@reduz Indeed, 32 bit and 64 bit compile and work for my VS2013 Community with env["x86_opt_vc"]=False

Thank you for asking that, I was too busy "figuring it out" to realise that i can test what i was suggesting you to do :)

But yeah, asm is strictly performance related code, that i know from those discussions in 2009 on the theora mailing list... so, turning it off will impact perf somewhat... but at least everybody should be able to compile...

**Comment by vnen** [comment]
I'm able to build both 32 and 64 with `env["x86_opt_vc"]=True`.

**Comment by punto-** [comment]
what's your compiler version? I can't build any assembler (32 or 64 bits)
with mine

$ cl --version
Microsoft (R) C/C++ Optimizing Compiler Version 18.00.40629 for x64

**Comment by vnen** [comment]
`Microsoft (R) C/C++ Optimizing Compiler Version 18.00.31010 for x86` (from `cl.exe`). Also works with VS2015, which is version 19.00.23026.

**Comment by Griefchief** [comment]
@vnen Since it should be confirmed at this point (I and punto) that (at least) community versions can't compile x32 at this point can you please open this issue. I'll do some testing in the meantime and maybe ask you to compile with some stuff on/off...

Just as a note for everybody, 

env["x86_opt_vc"]=False

enables my version of MSVC to compile both arhitectures... it's an option for stable... in the meantime, i'll try to learn more about this.

**Comment by vnen** [comment]
I'm with a clean env now and tomorrow will try the 2013 community version, since I only installed 2015 community, which still can build for 32-bits with current HEAD.

**Comment by Griefchief** [comment]
TL DR (I can compile latest 32 bit HEAD now... it's my fault)

@punto- check bellow, i think this might be your problem as well:

@vnen I think you might not have to :)

I think i figured out what my problem was (still compiling... yep, i just compiled a 32bit version). I was using "VS2013 x64 Native Tools Command Prompt" shortcut because i could swear that "Developer Command Prompt for VS 2013" didn't exist in the folder when i tried to find it.

In fact, i THINK i might have added it manually earlier via the accepted answer from this SO question:

http://stackoverflow.com/questions/21476588/where-is-developer-command-prompt-for-vs2013#22871532

But i just failed to use it then and used a shortcut on my taskbar to the 

"VS2013 x64 Native Tools Command Prompt"

instead (and with some kind of AMD setting on top of that... i guess those amd tools don't like _asm... they are in "Program Files(x86)\ Microsoft Visual Studio 12.0\VC\bin\amd64)...


@punto- Can you please check if you have the same problem I did? Can you use the first link to stack overflow to build a proper developer prompt if you do not have one in:

C:\Program Files (x86)\Microsoft Visual Studio 12.0\Common7\Tools\Shortcuts

Though, on closer inspection, the dev command prompt shortcut was probably created during the instalation (or 2 minutes after it) based on the create dates of all of the shortcuts in that folder... so... 

@punto- , do you have the Developer Command Prompt for VS2013" shortcut in:

C:\Program Files (x86)\Microsoft Visual Studio 12.0\Common7\Tools\Shortcuts

and can you build after this info?

**Comment by Griefchief** [comment]
And just to add more info (for other people that stumble on this issue as well), this is why it kept failing:

I kept building with 64 bit vc crosscompiler (in the amd64 folder) to build 32bit version of godot... and 64 bit version of the VS compiler/crosscompiler simply does NOT like ANY form of _asm (or __asm, which IS in fact supported for VC (32 bit compiler) and is present in the x86_vc folder of theora (x86 folder is for gcc))... 

http://stackoverflow.com/questions/27246382/what-is-the-difference-between-x64-native-tools-command-prompt-and-x64-cross-too

Sorry for the SPAM folks, but I used this issue to get my feet wet with godots code and scons build files for theora, I expect now to contribute something other then bug reports soon :)

**Comment by Griefchief** [comment]
It seems there are (at least) 4 diff vc compilers for x86/x64 located in these folders

```
%VCINSTALLDIR%\bin\cl.exe  (this would be x32 standard compiler)
%VCINSTALLDIR%\bin\amd64\cl.exe (this is vc 64 bit compiler...)
%VCINSTALLDIR%\bin\amd64_86\cl.exe (i think this is the 64bit crosscompiler for 32bit builds, it gives same info as standard compiler \bin\cl.exe)
(I don't know where the 32bit to 64bit crosscompiler is...)
```
I've studied how environent .bat files setup the compiler environments (how they choose between different versions via Dev Tools/Native/Cross Compile tools)... there are no additional ENV vars that may help figure out what actual VC compiler is being called... but...

They way these compilers are "switched/called/setup" (via Dev Console and Native/Cross Comiple tools) is by .bat files that add to the %PATH% variable... they just add the directory of the compiler to the cl.exe into %PATH% var... thet is how bats add the compiler into the path... (and Dev Console and Native/Cross Compile consoles just call these .bat files)...

So, i've looked at SConstruct and platform/windows/detect.py i think it might be possible to detect what actuall compiler is being called via the %PATH% var described above and i think I can write the code that can warn the user that the compilation might fail when amd64 or amd64_86 compiler is used...

So, reduz, it's probably something you don't want in SConstruct but something like bellow can probably detect the version of the vc compiler and warn the user... and i can build and test this if you happen to want this... I'm putting this here to let you know in case you might need this info later or similar... not a biggie for me, i looked into this for practice.

````Python
# example code = pseudo python/actionscript3 amalgation, bear with me please :)

# Test if standard x64 compiler will be used via string search
vc_amd64_compiler_detection_index =  env['PATH'].indexOf(env['VCINSTALLDIR']+"\bin\amd64;");
vc_amd68_x86_compiler_detection_index = env['PATH'].indexOf(env['VCINSTALLDIR']+"\bin\amd64_86;");

if(vc_amd64_compiler_detection_index > vc_amd68_x86_compiler_detection_index):
       vc_amd64_chosen_compiler =  vc_amd64_compiler_detection_index
else:
       vc_amd64_chosen_compiler =  vc_amd68_x86_compiler_detection_index
# Check that amd64 is indeed the first cl to be called by checking if the standard 32 bit compiler path
# is *behind* the amd64 compiler path
vc_standard_compiler_path_index = env['PATH'].indexOf(env['VCINSTALLDIR']+"\bin\;");
if(env['bits']='32' 
   AND vc_standard_compiler_index > 0
   AND vc_amd64_chosen_compiler > 0
   AND vc_standard_compiler_index > vc_amd64_chosen_compiler
   AND env['force_vc_amd64_crosscompilation'] = 'no'):

      print("BURN/CRASH, you can't compile 32bit version of godot with "+
"Visual Studio 64 bit compiler/crosscompiler because it does not support assembly compilation..."+
" without disabling theora (theora=no) or other stuff that might use assembly."+
" Also, you might be using 'VS 20xx Native Tools Command Prompt' or"+
"'Cross Compile Command Prompt'. Switch the the Developer Console "+
"(or x86 Native Tools Command Prompt) if that is the case.")
````

**Comment by vnen** [comment]
@Griefchief glad you figured this out. Can you build for 64-bits using the correct Developer prompt if you set `env["x86_pt_vc"]=True`? Maybe the fix made by @reduz was not even needed. If those are optimizations, they're better to be enabled.

On thing I didn't notice before is that's @punto-'s compiler version show `for x64` while mine had `for x86`, so that might be same thing @Griefchief just described.

**Comment by Griefchief** [comment]
@reduz @vnen Yes, indeed, i can build 64 bit with the 32 bit crosscompiler... I can't build any _asm (or __asm) with the 64bit crosscompiler to 32bit...

I don't see where puntos answer with his version of cl.exe but i think i too remmeber it being x64 and i know from yesterday that that is indeed BOTH the amd64 compiler and amd64_86 (and both refuse to build _asm)

Well, this is embarrasing :)

Anyway, we should probably note somewhere (and in tuts for people) at this point that only VC 32 bit compiler/crosscompiler is supported by godot at this point unless you would like some "ugly" compiler detection code like the sample one i wrote 2 lines ago...

We still have an "issue" at this point with the "fix":

1) 32 bit compiler can build both 32 and 64 bit via the crosscompiler

2) 64 bit can ONLY build 64 bits with the current line, not 32bit:  
```
env["x86_opt_vc"]=env["bits"]!="64"
```
And that is because we disable assembly (_asm and _ _asm) when compiling for 64 but not for 32...

**Comment by vnen** [comment]
I don't see what is the point of this "crosscompiler". The compiler for x86 (the one in the Dev prompt) can build for both architectures anyway. What's the benefit of a 64-bits compiler? It's not multi-core, since you can just pass "`-j 4`" to Scons and build using 4 threads.

We *could* add a workaround for people using the wrong prompt, but it's mostly a matter of documentation. Even the old wiki already says that Native Tools is the wrong on (though it's note in the middle of the page).

**Comment by Griefchief** [comment]
The x86 "compiler" (actually 3 diff compilers) is capable of compiling for 3 different arhitectures... when x86 compiler compiles a x86 executable, it's called native compiling (i think). When a x86 compiler compiles a x64 executable, it's called crosscompiling (I think you folks also call compiling linux -> windows or windows -> crosscompiling)... Crosscompilers exist so that you can compile for all arhitectures without having the native OS...

Anyway, these 3 diff x86 are in these folders...
```
VC/bin
VC/bin/x86_amd64 (i think this crosscompiles for the x64 arhitecture)
VC/bin/x86_arm (i think this one crosscompiles for ARM, but we don't care about that currently)...
```
So, that's what a crosscompiler does (it compiles the same source for a different arhitecture then the one it is written in). A transcompiler takes one language source, creates a second languages source (say, from C to Javascript) and then compiles for whatever arhitecture... If i had to guess, Emscripten is a transcompiler...

Anyway, you are correct, the 64-bits compiler(and transcompiler) isn't that important, but because of this "fix" i called for and we now implemented in (b09a85d) we now also half support the x64 bit compiler. When we compile scons p=windows bits=64 it will now succeed because (b09a85d) disables theoras assembly for 64 bits... so... we half support now, but we are also breaking the 64bit build of the x86 compiler, because it's not using assembly...

So we need (b09a85d) reverted to True... because I am a newb... that or we bother to detect what exact compiler (cl.exe in what directory) is compiling and we turn of assembly for x64 versions...

I don't think there is an easy way (other then the %PATH% code i showcased a few comments ago) and you folks will probably not want that stinking up SConstruct so documentation is probably better...

Also, thanks for the link @vnen : https://gist.github.com/ZuBsPaCe/0e1c31960e84df33e20f

It's what enabled me to solve the issue when i went hunting through it, i didn't realise that there are 20 diff compilers in VC/bin...

**Comment by Griefchief** [comment]
So, @vnen, please restore [bug] until (b09a85d) is reverted 

(the reason is in the second half of the previous comment... "we are also breaking the 64bit build of the x86 compiler, because it's not using assembly for x64" when it can use it and compiles for x64 fine...).

not quite sure why but it does, i guess it's just not using the last 32bits of the registers or the optimining compiler can convert the asm to 64 bit... no idea

I'll see what i can do about docs.

**Comment by vnen** [comment]
Wasn't supposed to close, hehe.

Anyway, I reverted the fix, as it doesn't fix anything anyway. Will keep this open until proper documentation is online.

@Griefchief, if you (or someone else) wants to make the check for compiler and then disable optimizations (or fail gracefully), you're welcome.

**Comment by Griefchief** [comment]
OK, I will see if something similar can be done gracefully in SConstruct.

@vnen what permissions are needed for an account to edit this wiki page: 
http://godotengine.org/projects/godot-engine/wiki/Compiling_for_Windows

and where to get em? 
http://151.236.22.217/users/2734

**Comment by vnen** [comment]
@Griefchief Not sure if @TheoXD is still watching the thread: http://godotengine.org/topics/11664. Just hope he checks this issue after my ping. You might catch him on IRC or Facebook. (I have the permission to edit the wiki, but not to give it to other people).

**Comment by TheoXD** [comment]
@Griefchief, it would have been faster if you made a post in that thread, since I check it more frequently than github. Anyway now you can edit the wiki, if your nickname matches the one here on github.

**Comment by Griefchief** [comment]
@TheoXD No biggie, i'll be playing with the wiki after SConstruct from this
bugreport so i wasn't in a rush. Thank you for the perms.

On Mon, Feb 1, 2016 at 7:34 PM, TheoXD <notifications@github.com> wrote:

> @Griefchief <https://github.com/Griefchief>, it would have been faster if
> you made a post in that thread, since I check it more frequently than
> github. Anyway now you can edit the wiki, if your nickname matches the one
> here on github.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/godotengine/godot/issues/3098#issuecomment-178114264>.
>



-- 
Thanks,

Aleksandar Danilovic.

**Comment by punto-** [comment]
what was the status of this? It still happens on my compiler version
(cl.exe 18.00.40629)

**Comment by Griefchief** [comment]
Reverted because it's not a bug, theora/godot can't be compiled on a 64bit
version of the msvc compiler without assembly... Read above or read this
link on how to open up a 32bit version of the compiler (to compile both
versions of godot on it):

http://www.godotengine.org/projects/godot-engine/wiki/Compiling_for_Windows#Compiling

In the next few days, i'll try to add detection into sconstruct but i can't
currently because of work... i expect to be freer from tomorrow.

On Sun, Feb 14, 2016 at 3:55 AM, punto- <notifications@github.com> wrote:

> what was the status of this? It still happens on my compiler version
> (cl.exe 18.00.40629)
>
> On 2 February 2016 at 14:03, Aleksandar Danilovic <
> notifications@github.com>
> wrote:
>
> > @TheoXD No biggie, i'll be playing with the wiki after SConstruct from
> this
> > bugreport so i wasn't in a rush. Thank you for the perms.
> >
> > On Mon, Feb 1, 2016 at 7:34 PM, TheoXD <notifications@github.com> wrote:
> >
> > > @Griefchief <https://github.com/Griefchief>, it would have been faster
> > if
> > > you made a post in that thread, since I check it more frequently than
> > > github. Anyway now you can edit the wiki, if your nickname matches the
> > one
> > > here on github.
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <
> https://github.com/godotengine/godot/issues/3098#issuecomment-178114264
> > >.
> > >
> >
> >
> >
> > --
> > Thanks,
> >
> > Aleksandar Danilovic.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/godotengine/godot/issues/3098#issuecomment-178689241
> >.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/godotengine/godot/issues/3098#issuecomment-183807537>.
>



-- 
Thanks,

Aleksandar Danilovic.

**Comment by punto-** [comment]
It looks like the build command for 32 and 64 bits with msvc is exactly the
same.. The 32 bit compiler can cross compile to 64? What's the compiler
flag?

On 14 February 2016 at 04:52, Aleksandar Danilovic <notifications@github.com
> wrote:

> Reverted because it's not a bug, theora/godot can't be compiled on a 64bit
> version of the msvc compiler without assembly... Read above or read this
> link on how to open up a 32bit version of the compiler (to compile both
> versions of godot on it):
>
>
> http://www.godotengine.org/projects/godot-engine/wiki/Compiling_for_Windows#Compiling
>
> In the next few days, i'll try to add detection into sconstruct but i can't
> currently because of work... i expect to be freer from tomorrow.
>

**Comment by Griefchief** [comment]
Lolz

Great catch punto! I've been messing around with this today finally, the last 4-5 hours, and here is what i can tell:

punto, there is no compiler flag for which architecture MS VC should compile to, it is controlled by **PATH** (i.e it depends on what console you fire up and call scons with (**Dev, Native, CrossCompile** ones)...

(More info on the lack of a compiler flag)
[http://stackoverflow.com/questions/2548138/how-to-compile-x64-code-with-visual-studio-in-command-line]

-----

The consequence is that, when compiling via VS VC compiler, it does not matter what you set the bits param in scons, for example:

``scons p=windows bits=64 ``

has no actual effect (at all). If you do this via the **Developer Command Prompt**, it will compile the **32bit** version regardless (**PATH** is set to the 32bit native compiler).


If you call scons via **VS2013 x64 Native Tools** it will compile a **64 bit** version even if you do:

``scons p=windows bits=32``

The good news is that MS VC x64 native does seem to compile and run godot properly (tested with theora=no so far, but i expect it to run with it as well, if assembly is turned off).

-----

At this point, we need to:

1. Update the documentation to explain the differences between the different compilation prompts (Dev, Native, CrossCompile), i.e. that **bits** parameter does nothing in SCons when compiled with visual studio compilers (MS VC)

2. Decide how to handle this: 

- My original idea was that we can force all visual studio compilation attempts to set bits=prompt_controlled or similar... and abort with an explanation if the user does not set it that explains what different prompts/compilers do... and there is also the bellow idea, but i don't think it's the correct one anymore... and the reason is that vsvars.bat sets up a lot of stuff that the different compiler versions expect... in essence, we should to do this "bits disable", i can't just setup the PATH like i wanted to bellow... it won't work, and we are removing control from the user (setting up a 32bit compiler path on x64 bit Native Command prompt, for example).

- (I wanted to do this one but no, bad idea, here for reference, to better explain why it won't work) We should probably just add the correct compiler to the SCons internal PATH variable and attempt to compile with the propper version of the compiler... (but nope, vcvarsall.bat sets up a lot of env variables depending on the wanted compiler, i looked at it a few weeks ago, just no, won't work).

-----

So, @reduz, @vnen and any other interested devs, can we do something like this:

``scons p=windows bits=prompt_controlled``

and if bits is not set to "prompt_controlled", to write something like bellow and abort:

``
Aborting: Canot set bits parameter for the Visual Studio VC compiler. You are required to call the correct prompt (compiler) yourself in order to have it compile godot for the correct arhitecture. The options are as follows:

For a 32 bit binary:
Developer Command Prompt for VS20xx    (VsDevCmd.bat)
VS20xx x64 Cross Tools Command Prompt (vcvarsall.bat amd64_x86)

For a 64 bit binary:
VS20xx x64 Native Tools Command Prompt (vcvarsall.bat amd64)
VS20xx x86 Cross Tools Command Prompt (vcvarsall.bat x86_amd64)

After firring up the correct prompt you want, set the bits parameter to bits=prompt_controlled to signify that you understand how to compile different arhitectures via Visual Studio, and to be able to compile.
``

Something like that... or whatever you wish. sorry for the long read, and let me know if you want me to do a detection script like this with bits=prompt_controlled (we really should let the user fire up the correct prompt himself)

Or however you want this handled...

**Comment by Griefchief** [comment]
And just to note, as i understand above will be a hard read...

I do believe we should "disable" the bits parameter of scons for Visual Studio compilation and do something like bits=prompt_controlled...

I thought about how to do it otherwise, but at the end of the day, the user simply has to do it, because it's both too complicated for scons (almost impossible), and wrong to reset the users prompt, even if he doesn't understand the choices... we should just teach them.)

``scons p=windows``

can just work, i do not wish to disable this option

Thank you.

**Comment by punto-** [comment]
I agree with disabling the bits parameter, but we should detect the
architecture of the prompt and add it to the binary name, otherwise we're
left with "godot.windows.opt.exe", "godot.windows.opt.32.exe" and
"godot.windows.opt.64.exe", wtf is the first one? Also if we know the bits,
we'll know to disable the theora optimizations.

So is there something in the environment that tells us the architecture?

On 21 February 2016 at 19:12, Aleksandar Danilovic <notifications@github.com

**Comment by Griefchief** [comment]
The only propper way that i am aware of would be to check PATH(just reading it, not us touching it)... that would give us enough info to setup the name...

 something similar to the detection code i created way back in one of the upper comments... if i find something else, i'll report...

Alternatively, we can just let it create a name like it does with the first one (with the first one it just does not setup the SUFFIX properly... i mean, it skips over the bits=="32" and bits=="64" check in SConstruct that adds the numbers to the SUFFIX env variable...

**Comment by vnen** [comment]
@Griefchief I don't have much spare time to check this, but I want to thank you for investing time on this. I had no idea that it was compiling 32bits even with `bits=64` and it seems nobody else had, so this is a great find.

I think it is good to omit the bits parameter to proper build and warn the user if it is set (or maybe only if it's set wrongly). Also a line with a warning would be good (like `Using detected 32bits compiler` or something worded better).

**Comment by Griefchief** [comment]
You're welcome, i wanted to work on godot source in some way or another, so i might as well finish what i started here... not awesomely fun but i also never worked with build systems so it was cool in the end... punto kind of beat me to it though, i woudn't realise this issue if he hadn't said something about it. so thanks to punto again :)

That said, the compiling issue is def there, no need for you to check, i've confirmed it in too many different ways (like hex editor checking the executable, as well as looking at it from windows manager, it says it there etc).

Yep, i'll work on just warning and aborting if bits is set, no need for bits=prompt_controlled stuff... and some detection as well, punto is right that we need to disable optimisations.

**Comment by Marqin** [comment]
This bug is still in `HEAD`!

https://ci.appveyor.com/project/Marqin/godot/build/job/uwubv3my5kxjqp66

**Comment by Griefchief** [comment]
Indeed, it is.

Currently, the revert that we did way back has put it back in (though the fix, in retrospect, never really fixed it, but it at least allowed to compile by controlling the bits parameter). That said, i did work on the detection code a few days ago and just wanted to continue today... so, heh.

I am hoping to have the detection code done within a day or two.

In the meantime, @Marqin I noticed something about the build you are doing (and not sure if you read the whole wall of text so i'll note again):
```
VS=C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat, BITS=32, TOOLS=yes
VS=C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat, BITS=64, TOOLS=yes
```
This should always give you the same arhitecture... You can not control VC via the bits parameter, because different cl.exe files compile for different arhitectures (there is no bits flag in vc compiler, so scons bits is a bit useless...).

Something like this is correct:

```
VS=C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat, TOOLS=yes
VS=%comspec% /k ""C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\vcvarsall.bat"" amd64, TOOLS=yes
```
//note: I have no clue what %comspec% /k stuff does, i just copied it from the shortcut here in my win7... please investigate it, might not be important...

Notice how the bits param is mising... i see though that you are using it in your pythong script... you can leave it, just note that it won't do anything currently... other then create a suffix in the filename for the build so that you can recognize it visually or if the files happen to be in the same folder...

And cl.exe (visual c) will itself build the arhitecture it is designed to build... regardless of what you set the bits parametetar to....

This vcvarsall file setups the path to the correct cl.exe and lots of other things cl.exe needs in it's environment.

I hope this made sense, and sorry if you realised this earlier, i just wanted to save you some headache. To repeat again more simply, I believe you built 4 32bit arhitectures with your current scripts and build jobs. I don't see it complaining about _asm in your builds so that must mean that you built 4 diff 32 bit versions... Once it complains about _asm, you are buliding a 64 bit built... otherwise, you are not.

I'll try to finish the detection code I started in a day or two (I hope PATH detection will work, it should).

As a note, Juans fix from way back (that we reverted cause we still didn't know what exatcly it did) will allow _asm to be turned off if you manually type in bits=64... so it will allow compilation on a 64bit arhitecture, currently there should be no way to build for 64 bit cause assembly should always be on... So, that Juans fix will allow compilation, but it doesn't actually controll the arhitecture (vcvarsall.bat will, as I explained above).

https://github.com/godotengine/godot/commit/ac9807e89837642004f9e0f07fd306038e0b82f6

So, that capability is there to allow a 64bit build to actually build. It's not a fix per se, as you will realise if you read carefully the whole exchange above, there is no "real" fix, we can only detect and warn, and setup some stuff through that detection, which will allow builds to go through.

I just noted that while I am working on a propper detection code if you the builds to work asap...

I'll go code now.
Thank you.

**Comment by Marqin** [comment]
@Griefchief , it doesn't build with amd64, and Godot docs say that using bits in scons will work with MSVC - http://docs.godotengine.org/en/latest/reference/compiling_for_windows.html#note-for-godot-2-0-if-you-are-having-issues

**Comment by Marqin** [comment]
@Griefchief I don't understand your command with that comspec, what I do here is:

    VS: C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat

and later:

    if defined VS call "%VS%" x86

Because with `amd64` it doesn't compile theora and glew drivers AND Godot Docs say that we should use x86. Maybe docs lie.

**Comment by Marqin** [comment]
@Griefchief, here's the bug I got in when compiling with MSVC 14 x64 - #4113 
and with 12 x64 it failed on that _asm in theora.

**Comment by Griefchief** [comment]
Docs do not lie, x86 is currently only supported one (technically) because of this bug...

That said, you can compile for x64, but you have to manually fix this issue here we are commenting on, so automated builds won't work with x64 just yet, sorry...

Long story short is that 64bit MSVC does not support 32 bit assembly... it can't compile it. So you get theora ASM.

Why the issue you discovered happens (ptrdiff_t), I have no idea... it does not seem related to assembly discussed here, but a separate issue... that said, i'm not a C/C++ expert...

But it does seem we have 2 separate issues... I'll work on fixing this one for now, it's time to close this one.

**Comment by Marqin** [comment]
@Griefchief , but you said

>And cl.exe (visual c) will itself build the arhitecture it is designed to build... 

And Godot docs say:
>One more thing, 32 bit compiler can compile both 32 bit Godot and 64 bit Godot. 

So now I'm confused - which one is true?

**Comment by Griefchief** [comment]
I wrote that "One more thing", a good chuck of that compiling paragraph, when i had no clue how visual c compiler worked (in my defense, i don't think anybody developing godot knew how it builds different arhitectures)... So, the docs are wrong i guess :)

I need to rewrite them as part of fixing this bug...

If you reread this bug here we are commenting on, you'll notice that we stumble in the dark for months until i figured out how visual c worked one day with puntos help :)

So, what you are doing, is correct, you understand how vcvarsall.bat works... when you do this:

```if defined VS call "%VS%" x86```

So, you are calling vcvars bat file properly... i didn't notice this call in the python file when i looked over it...

So, docs are wrong...

**Comment by Marqin** [comment]
I just found "bug" in Scons files, but I have to check if it helps.

**Comment by Marqin** [comment]
After fix ( using /DWIN64 instead of /DWIN32 ):
Good news is that is doesn't break on theora. Bad news is, because it breaks before, on Opus...

I also found fix for the other bug, it was because Godot's `glew.h` was changed. As for Opus - we seem to have very old version of it..

**Comment by Marqin** [comment]
And there is nowhere written which one.
@reduz @punto- which version of Opus Godot have?

**Comment by Marqin** [comment]
Disabled opus and it still errors on Theora :disappointed:

**Comment by vnen** [comment]
@Marqin it's 1.1, I believe, as I added it before 1.1.1 was release. And opusfile v0.6. Still haven't got the time to update and test in all platforms.

**Comment by akien-mga** [comment]
@Griefchief Did your PR fix this issue too?

**Comment by Griefchief** [comment]
Yes, i believe you can close this issue.

**Comment by Griefchief** [comment]
@akien-mga Yes, I believe you can close this issue.
